blues/airnote.live
# bug: fix flaky dashboard charts & remove PIN requirements to view settings information

# Problem Context

This PR tackles three main problems:

1. Intermittently the dashboard fails to load and gives a white screen of death, which I believe is due to it attempting to render charts before there's data back from Notehub to render the charts with. 
2. The `Unable to fetch device details` error shows in the Settings page when the user has an invalid PIN for the current Airnote device UID. This is confusing and incorrect, if the PIN's invalid, we should show them an indicator of that but still let them view individual Airnote data, just not update it.
3. If a user has an Airnote's device ID, the Settings and Dashboard links should always be visible in the header even if a user has no PIN - they should just not be allowed to be updated unless the correct PIN for a particular Airnote is present.

## Changes

- Fix up all the charts to no longer try to render before their readings from Notehub have been fetched, and add more error handling to handle it if data fails to come back.
- Decouple main dashboard from relying on both device ID and PIN to view the `Settings` and `Dashboard` links in the nav bar (now a user must only have a device ID to be able to see all aspects of an Airnote)
- Consolidate all error messaging for the app into a set of constants and a function to render the appropriate HTML message based on the `errorType` passed in.
- Decouple settings page from depending on a valid PIN to display Airnote device details. Now, the details are displayed regardless of PIN status, but the input boxes are left disabled if the PIN is missing entirely or invalid and an appropriate error message is displayed (I'm still working on the exact messaging with Michelle)
- Move all API calls in the app to the `server.js` file instead of letting any originate out of the client, and use the `device.js` file as the routing connection between the client and server.
- Fix broken unit tests and add a few more for the increased functionality of the `Settings.svelte` file and `device.js` file.

## Screenshot (if applicable)

Valid PIN no longer required to see Airnote device info (but inputs disabled if that's the case)

![Screen Shot 2022-08-26 at 5 06 16 PM](https://user-images.githubusercontent.com/20400845/186990768-a9fccac5-0a6f-4414-b7c9-26901baa01cd.png)

## Testing

Unit / integration / e2e tests?

Additional unit tests added.

Steps to test manually?

To test if the navbar links render in the browser without a pin required,
1.  Open an incognito browser window and go to: http://localhost:5555/dev:864475044215258/dashboard
2. You should land on the dashboard of an Airnote and be able to see the nav bar links
3. Click the "settings" link
4. Observe you can see the settings page and all the Airnote settings, but the URL has a query param of `&pin=` that is empty and there's an error message stating the PIN is missing or invalid and so users can look but not change settings for this Airnote.

To test if the dashboard always renders correctly and never gives a white screen of death:
1. Go to: http://localhost:5555/dev:864475044215258/dashboard
2. Refresh the page several times and see the dashboard continue to render correctly every time, no intermittent errors happening where the screen goes blank except for header and footer.

Any other sorts of testing notes to include?

I'm still working out the final wording with Michelle for the error messaging on the Settings page when users have a missing or invalid PIN or fail to save their device changes to Notehub.

## Any other related PRs

n/a

## Ticket(s)

https://trello.com/c/fktDJ3CT/7-handle-ui-bugs-where-the-page-breaks-when-the-data-from-notehub-is-not-yet-present-or-user-has-invalid-pin-when-hitting-the-sett

https://trello.com/c/w09Pzulr/6-make-settings-and-dashboards-links-always-visible-in-airnote-header
